doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,176 @@
+# Materialize Formalism
+
+Materialize is a system that maintains views over changing data.
+
+At Materialize's heart are *time-varying collections* (TVCs).
+These are multisets of typed *data* whose contents may vary arbitrarily as function of some ordered *time*.
+Materialize records, transforms, and presents the contents of time varying collections.
+
+Materialize's primary function is to provide the contents of time-varying collection at specific times with absolute certainty.
+
+# Time-varying collections
+
+Materialize represents TVCs as a set of "updates" of the form `(data, time, diff)`.
+The difference `diff` represents the change in the multiplicity of `data` as we go from `time-1` to `time`.
+To go from a sequence of collections to this representation one subtracts the old from new frequencies for each `data`.
+To go from this representation to a sequence of collections, one updates the multiplicities of each `data` by `diff` for each `time`.
+
+Materialize most often only has partial information about TVCs.
+A TVC's future is usually not yet completely known, and its deep past also may not be faithfully recorded.
+To represent what Materialize knows with certainty, Materialize maintains two "frontiers" for each TVC.
+A frontier is a lower bound on times, and is used to describe the set of times greater or equal to the lower bound.
+The two frontiers Materialize maintains for each TVC are:
+
+*   The `since` frontier, which communicates that for times `time` greater or equal to `since`, the accumulated collection at `time` is precise.
+
+    Informally, updates at times not greater than the `since` frontier may have been consolidated with other updates.
+    They are not discarded, but their times may have been advanced to the `since` frontier to allow cancelation and consolidation of updates to occur.
+    This is analogous to "version vacuuming" in multiversioned databases.
+    As the system operates, the `since` frontier may advance to allow compaction of data.
+    This happens subject to constraints, and only when correctness permits.
+
+    We think of `since` as the "read frontier": times greater or equal to it may still be correctly read from the TVC.
```

Comment:
nit: I think more precisely this is: 

> times greater or equal to it but not greater or equal to `upper` may still be correctly read from the TVC.

if you want to include this, you might swap the order of the `lower` and `upper` paragraphs.

Reply:
I changed `still` to `yet` hoping to get at the intended difference. The `since` guarantee is that the when a time comes, the updates at that time will be correct.